### ISSUE TITLE
Mac Web Engine timing out on Safari on Xcode 13.*

### ISSUE THREAD

**Issue by jmagman** [flutter/flutter#99885]
Seen in https://github.com/flutter/engine/pull/31915
```  
INFO: Simulator 88E8A88A-27B8-4CB4-92F5-C0A7D29C9EB9 booted.
...
  [31mFailed to load "test/paragraph_builder_test.dart":[0m
  Failed to run Safari iOS: Exception: Safari iOS failed with exit code 60.
  Standard output:
  An error was encountered processing the command (domain=NSPOSIXErrorDomain, code=60):
  The operation couldn’t be completed. Operation timed out
  Operation timed out
  .
  dev/browser.dart 143:9  new Browser.<fn>
```
https://logs.chromium.org/logs/flutter/buildbucket/cr-buildbucket/8820126131592201793/+/u/test:_felt_ios-safari_test/stdout

Simulator successfully booted, timed out waiting for Safari to launch.

**Comment by jmagman** [comment]
Similar to https://github.com/flutter/flutter/issues/98278 but Safari.
Maybe something in https://github.com/flutter/engine/blob/main/lib/web_ui/dev/browser_lock.yaml needs to be updated?

@yjbanov @ditman

**Comment by jmagman** [comment]
Also https://github.com/flutter/engine/blob/bf2bf3a3fd2b1a55f429884fef86d05d05156be2/lib/web_ui/dev/browser_lock.yaml#L19 probably needs to be upped to 98/98.1?

**Comment by jmagman** [comment]
This test passed in January on the same Xcode bundle/iOS simulator runtime bundles/macOS 12 https://ci.chromium.org/raw/build/logs.chromium.org/flutter/led/magder_google.com/9a0597c302a0ba1bf2edb738c803dbdcad105f47a13c5eec62f43f3a5e92cf16/+/build.proto?server=chromium-swarm.appspot.com

**Comment by yjbanov** [comment]
This is Safari for iOS, which is launched in an iOS Simulator. And Xcode and/or macOS may affect it. The Chromium codepath is sufficiently different that I don't think the two are related. Safari iOS configuration is specified [here](https://github.com/flutter/engine/blob/bf2bf3a3fd2b1a55f429884fef86d05d05156be2/lib/web_ui/dev/browser_lock.yaml#L29). Code that manages Safari on iOS is [here](https://github.com/flutter/engine/blob/main/lib/web_ui/dev/safari_ios.dart).

**Comment by jmagman** [comment]
> The Chromium codepath is sufficiently different that I don't think the two are related. 

Oh for sure was just a stray comment that it also may need to be bumped.

**Comment by jmagman** [comment]
> Safari iOS configuration is specified [here](https://github.com/flutter/engine/blob/bf2bf3a3fd2b1a55f429884fef86d05d05156be2/lib/web_ui/dev/browser_lock.yaml#L29). Code that manages Safari on iOS is [here](https://github.com/flutter/engine/blob/main/lib/web_ui/dev/safari_ios.dart).

Right, it's using iOS 13 simulator, which is available, and is being booted.  This test was passing on the same Xcode and runtime bundles in January, though the macOS version was 12.1 instead of 12.2 on the failing machines, not sure if that's related.

**Comment by jmagman** [comment]
Tried on the iOS 15 simulator, same results
https://logs.chromium.org/logs/flutter/buildbucket/cr-buildbucket/8820119593192648065/+/u/test:_felt_ios-safari_test/stdout

**Comment by jmagman** [comment]
@eyebrowsoffire investigated, it reproduced on Xcode 13 trying with different iOS simulator runtimes, but did not reproduce on Xcode 13.2.1.  I'm going to prep an Xcode 13.1 package to see if it repros there.  If it works then I'll bump the framework package, min Xcode tool version requirement, then circle back to the engine.

**Comment by jmagman** [comment]
Same thing on 13.1 https://github.com/flutter/engine/pull/32254
```
  Failed to run Safari iOS: Exception: Safari iOS failed with exit code 60.
```
https://logs.chromium.org/logs/flutter/buildbucket/cr-buildbucket/8818757082941132785/+/u/test:_felt_ios-safari_test/stdout

This used the previously uploaded iOS 15 simulator runtime since they have identical Info.plists and are both for the iOS 15.0 SDK.  But all the binaries actually differ from the [old](https://chrome-infra-packages.appspot.com/p/flutter_internal/ios/xcode/ios_runtime/+/xbExNWHTJfpNf8vDNXhmLNx9sjfFvAl4ed74gTrXl3cC) and [new](https://chrome-infra-packages.appspot.com/p/flutter_internal/ios/xcode/ios_runtime/+/bSZkcV-gse3F7uV37terlLhxOfepuLZVlOOQx6082KEC).  Need to figure out how to get the `ios-15-0` refs and tags from one to the other without potentially breaking the currently running CI to test if the new version fixes it (and if so, can we use the new runtime on Xcode 13 to avoid needing to [bump the tool requirement](https://github.com/flutter/flutter/pull/97746) again?)
Infra team advice on this would be much appreciated @godofredoc

**Comment by godofredoc** [comment]
Just moved the refs from the old to the new. If something fails we can move them back. The tags are the ones that should not be changed and impact the caches but refs are safe to move back and forth.

**Comment by jmagman** [comment]
> Just moved the refs from the old to the new. If something fails we can move them back.

@godofredoc Can you clear the `osx_sdk_13a1030d_13_1_15` and `osx_sdk_13a233_13_15` caches off the try builders to see if it works? 
https://github.com/flutter/engine/pull/32254/files#diff-9e73abde9fff445f31e5928e80d2d3f88847e02fd8f76d46f66f147fdc82786dR61
https://github.com/flutter/engine/pull/31915/files#diff-9e73abde9fff445f31e5928e80d2d3f88847e02fd8f76d46f66f147fdc82786dR60

**Comment by godofredoc** [comment]
Any particular builders that you would like to target? going through all of them requires some time

you can use `led -d id=<builderid>` to target a specific one

**Comment by jmagman** [comment]
Trying:
https://ci.chromium.org/swarming/task/59dae602cfa11b10?server=chromium-swarm.appspot.com

**Comment by jmagman** [comment]
It correctly picked the new `ios_runtime`, it's still running
https://logs.chromium.org/logs/flutter/led/magder_google.com/2db8442891a45ed4446c8e8eacaeb6a00d239e9cee02e3aab13417fec903e386/+/u/install_xcode/stdout
```
Packages:
  flutter_internal/ios/xcode/ios_runtime:bSZkcV-gse3F7uV37terlLhxOfepuLZVlOOQx6082KEC
```

**Comment by jmagman** [comment]
13.1 Xcode and newer iOS 15 simulator binary also timed out, failing to launch Safari https://logs.chromium.org/logs/flutter/led/magder_google.com/2db8442891a45ed4446c8e8eacaeb6a00d239e9cee02e3aab13417fec903e386/+/u/test:_felt_ios-safari_test/stdout

**Comment by jmagman** [comment]
@godofredoc it's not 100% but still seeing some timeouts:
https://ci.chromium.org/p/flutter/builders/prod/Mac%20Web%20Engine/9710
https://ci.chromium.org/p/flutter/builders/prod/Mac%20Web%20Engine/9716

**Comment by jmagman** [comment]
Revert https://github.com/flutter/engine/pull/32632

**Comment by zanderso** [comment]
I am increasing the priority of this because the 8 machines that are stuck on the old version of mac to support Mac Web Engine are causing capacity issues, e.g. sitting idle while other queues are filling up. Options for addressing this range from debugging the timeout, retrying after a tighter timeout, or disabling the tests on Safari while investigating offline.

**Comment by godofredoc** [comment]
I believe we hit a dead end on this issue, as far I can tell from the several weeks of experimentation the new versions of mac os / xcode can not open safari over ssh.

**Comment by godofredoc** [comment]
Would it be possible use chrome rather than safari for this test?

**Comment by yjbanov** [comment]
We cannot use Chrome. Tests run in Safari to protect ourselves from Safari-specific bugs. However, if we could consolidate all tests into desktop Safari. We only use iOS Safari for golden testing because the protocol the test harness uses for desktop Safari doesn't support taking screenshots. If we can use a method that supports screenshots, such as `safaridriver`, we could turn off iOS Safari tests, at least until we find a real need for it and have better infrastructure for testing on it.

**Comment by godofredoc** [comment]
There are new ideas on how to fix this issue. Safari on mac 12 added a new option to enable remote automation.

Configure Safari to Enable WebDriver Support
Safari’s WebDriver support for developers is turned off by default. How you enable it depends on your operating system.
High Sierra and later:
* Run `safaridriver --enable` once. (If you’re upgrading from a previous macOS release, you may need to use sudo.)
Sierra and earlier:
* If you haven’t already done so, make the Develop menu available. Choose Safari > Preferences, and on the Advanced tab, select “Show Develop menu in menu bar.” For details, see Safari Help.
* Choose Develop > Allow Remote Automation.
* Authorize safaridriver to launch the XPC service that hosts the local web server. To permit this, manually * run /usr/bin/safaridriver once and follow the authentication prompt.

(copy & paste from https://developer.apple.com/documentation/webkit/testing_with_webdriver_in_safari)


@keyonghan can you please help us verify that manually enabling remote automation makes the test pass with one of the devicelab machines?

if it works we will need to ask chromium to enable that config on the mac machines.

**Comment by keyonghan** [comment]
(copied from internal chat)

Tried on a devicelab mac bot, with `allow remote automation` checked, and `safaridriver` enabled.

Also validated `safaridriver` works
```
/usr/bin/safaridriver --port=4444

curl  -d '{ "capabilities": { "caps": { "nativeEvents": false, "browserName": "safari", "version": "", "platform": "ANY" } } }'  http://localhost:4444/session

{"value":{"sessionId":"91E6AE3E-C36E-486E-9DD9-12E729AB2420","capabilities":{"safari:platformVersion":"12.0.1","safari:automaticInspection":false,"webkit:WebRTC":{"DisableICECandidateFiltering":false,"DisableInsecureMediaCapture":false},"browserVersion":"15.1","strictFileInteractability":false,"browserName":"Safari","safari:useSimulator":false,"safari:automaticProfiling":false,"safari:platformBuildVersion":"21A559","acceptInsecureCerts":false,"setWindowRect":true,"platformName":"macOS","safari:diagnose":false}}}
```

But the build timed out on step `test: Run tests on macOS Safari`: https://luci-milo.appspot.com/raw/build/logs.chromium.org/flutter/led/keyonghan_google.com/2770b00e2e6e841a2232fa493d9c2ebad6c33daf70db18b5a26ee9c3b55fb746/+/build.proto, without any `safari` window open in the bot while running the test.
```
Running steps 
/opt/s/w/ir/cache/gold/goldctl auth --work-dir /opt/s/w/ir/cache/builder/src/flutter/lib/web_ui/.dart_tool/skia_gold/temp --luci

00:00 [32m+0[0m: [1m[1;30mloading test/paragraph_builder_test.dart[0m[0m                                                                                                                                                     
00:00 [32m+0[0m[31m -1[0m: [1m[1;30mloading test/paragraph_builder_test.dart[0m [1m[31m[E][0m[0m                                                                                                                                              
  Invalid argument(s) (onError): The error handler of Future.catchError must return a value of the future's type
  dart:async/future_impl.dart 181:7             _FutureListener.handleError
  dart:async/future_impl.dart 778:47            Future._propagateToListeners.handleError
  dart:async/future_impl.dart 799:13            Future._propagateToListeners
  dart:async/future_impl.dart 691:9             Future._propagateToListeners
  dart:async/future_impl.dart 574:5             Future._completeError
  dart:async/future_impl.dart 665:7             Future._asyncCompleteError.<fn>
  dart:async/zone.dart 1426:13                  _rootRun
  dart:async/zone.dart 1328:19                  _CustomZone.run
  dart:async/zone.dart 1236:7                   _CustomZone.runGuarded
  dart:async/zone.dart 1276:23                  _CustomZone.bindCallbackGuarded.<fn>
  dart:async/schedule_microtask.dart 40:21      _microtaskLoop
  dart:async/schedule_microtask.dart 49:5       _startMicrotaskLoop
  dart:isolate-patch/isolate_patch.dart 122:13  _runPendingImmediateCallback
  dart:isolate-patch/timer_impl.dart 405:11     _Timer._runTimers
  dart:isolate-patch/timer_impl.dart 429:5      _Timer._handleMessage
  dart:isolate-patch/isolate_patch.dart 192:12  _RawReceivePortImpl._handleMessage
  
  [31mFailed to load "test/paragraph_builder_test.dart":[0m Failed to run Safari macOS: Exception: Safari macOS failed with exit code -4..
  dev/browser.dart 143:9  new Browser.<fn>
```

**Comment by godofredoc** [comment]
@yjbanov should we remove the test while it is refactored to safaridriver? 

Context: enabling remote automation didn't help and made the problem worse not only failing for ios but also for mac.

**Comment by yjbanov** [comment]
Copying the offline discussion here:

Disabling these tests is risky as they are our only source golden tests for Safari. The best course of action is to expedite `safaridriver` support in Web Engine tests, so we don't need iOS simulators any more (and by extension remove our dependency on Xcode).

**Comment by godofredoc** [comment]
This is great news! please let us know if something will be needed for infra to support the new safari driver runner.

**Comment by yjbanov** [comment]
One thing that would be needed it to make sure safaridriver is enabled on all mac bots (more details [here](https://developer.apple.com/documentation/webkit/testing_with_webdriver_in_safari)). I'm looking for an owner on the web team to take this. If there are no volunteers I could take this next, after I'm done with my current P1.

**Comment by godofredoc** [comment]
@keyonghan can we ask the chromium labs team on the most efficient way to run that command in all the mac fleet?

**Comment by keyonghan** [comment]
> @keyonghan can we ask the chromium labs team on the most efficient way to run that command in all the mac fleet?

Tracked in crbug.com/1328969.

**Comment by keyonghan** [comment]
From the chromium bug, the command has been executed via puppet for all our chromium mac bots. Un-assigning myself here. Please reach out if any other action is needed from infra side.

**Comment by eyebrowsoffire** [comment]
This is in progress, I'm actively working on this.

**Comment by eyebrowsoffire** [comment]
I have a PR out with the tests based on safaridriver. However, at this point I'm blocked on (crbug.com/1328969) which Keyong mentioned above

**Comment by godofredoc** [comment]
Chromium team is working on enabling safari driver. Previously the command was run as root and it was not working for the chromebot user.

@keyonghan is following up with chromium team.

**Comment by keyonghan** [comment]
Configs are enforced.  @eyebrowsoffire please give it try.

**Comment by keyonghan** [comment]
https://github.com/flutter/flutter/issues/105721 was resolved via https://flutter-review.googlesource.com/c/recipes/+/31200.

Bumping xcode to 13 in: https://github.com/flutter/engine/pull/34051

**Comment by jmagman** [comment]
I believe this can be closed then?

**Comment by eyebrowsoffire** [comment]
I believe it can!
